subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -2443,4 +2438,42 @@ public Optional<? extends RuntimeException> mapFailure(String type, Collection<T
             return DefaultConfiguration.this.mapFailure(type, failures);
         }
     }
+
+    /**
+     * Helper class that allows all configuration usage information to be stored in a byte using bitwise operations.
+     * <p>
+     * This may slightly trim the memory usage of configuration instances, which can be numerous in larger builds.
+     */
+    private enum Usage {
+        CONSUMABLE((byte) 1),
+        RESOLVABLE((byte) 2),
+        DECLARABLE_AGAINST((byte) 4),
+        DEPRECATED_FOR_CONSUMPTION((byte) 8),
+        DEPRECATED_FOR_RESOLUTION((byte) 16),
+        DEPRECATED_FOR_DECLARATION_AGAINST((byte) 32);
+
+        private static final byte EVERY_NONDEPRECATED_USAGE = (byte) (CONSUMABLE.mask | RESOLVABLE.mask | DECLARABLE_AGAINST.mask);
+
+        private final byte mask;
+
+        Usage(byte mask) {
+            this.mask = mask;
+        }
+
+        public byte toggle(boolean allowed, byte currentUsage) {
```

Comment:
The name `toggle` makes me think this changes the current value to the opposite. Consider something like `setEnabled`

Reply:
I specifically didn't choose `setEnabled` because I didn't want to give the impression this was setting a field on the enum.  I see your point about `toggle` though.

How about `updateUsage`?

Reply:
Even just `update` or `set`

I'm fine with updateUsage